### PR TITLE
Better support for custom party hints for non-operator users

### PR DESCRIPTION
**Pull request by pasindutennage-da**
Fixes https://github.com/hyperledger-labs/splice/issues/3263

Related comment https://github.com/global-synchronizer-foundation/docs/discussions/26#discussioncomment-15209634

### CODE REVIEW

File: apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/admin/http/HttpValidatorHandler.scala [hunk 1]
```diff
@@ -50,6 +50,7 @@ class HttpValidatorHandler(
       .onboard(
         name,
         None,
+        Some(true),
```
Comment (pasindutennage-da) [inline apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/admin/http/HttpValidatorHandler.scala]: doesn't matter if this is true / false, a new party will be created

File: apps/validator/src/main/openapi/validator-internal.yaml [hunk 2]
```diff
@@ -551,11 +551,19 @@ components:
...
+          description: The name of the user to onboard.
         party_id:
           type: string
           description: |
-            The party id of the user to onboard. This is optional and if not provided
-            a fresh party id will be generated by the backend.
+            The party id of the user to onboard.
+            If no party_id is provided then a fresh party will be generated, using the 'name' as the Party Hint.
+            If party_id is provided and createPartyIfMissing is false, then the party must already exist on the ledger. The existing party will be assigned to the user.
+            If party_id is provided and createPartyIfMissing is true, then a new party will be created using the provided party_id as the Party Hint.
```
Comment (isegall-da) [inline apps/validator/src/main/openapi/validator-internal.yaml]:
```suggestion
            If party_id is provided and createPartyIfMissing is true, then: if a party with the provided party_id exists, the user will be associated with it. Otherwise, a new party will be created using the provided party_id, and the user will be associated with that new party.
```

(to make it more explicit, and also correcting "using the provided party_id as the Party Hint" which is incorrect. The provided ID is the full ID, it's not the hint).

Comment (pasindutennage-da) [reply apps/validator/src/main/openapi/validator-internal.yaml]: Resolved: Made the change

Comment (pasindutennage-da) [reply apps/validator/src/main/openapi/validator-internal.yaml]: Resolved: Made the change

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/ValidatorIntegrationTest.scala [hunk 3]
```diff
@@ -578,4 +578,56 @@ class ValidatorIntegrationTest extends IntegrationTest with WalletTestUtil {
...
+      aliceValidatorBackend.onboardUser(name, partyId, createIfMissing)
+    }
+
+    clue("Assign new user to existing Validator Party (party_id provided)") {
+      val assignedPartyId = onboard(
+        name = testUser1,
+        partyId = Some(aliceValidatorParty),
+      )
+      assignedPartyId shouldBe aliceValidatorParty
+      aliceValidatorBackend.listUsers() should contain(testUser1)
```
Comment (martinflorian-da) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/ValidatorIntegrationTest.scala:609]: nit: Shall we also add `aliceValidatorBackend.listUsers() should *not* contain(testUser1)` clauses before we do the `onboard`? For a tiny bit extra safety.

File: apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/util/ValidatorUtil.scala [hunk 4]
```diff
@@ -110,12 +111,22 @@ private[validator] object ValidatorUtil {
...
+            )
+          } else {
+            connection.createUserWithPrimaryParty(
+              endUserName,
+              party,
+              Seq(),
+            )
+          }
         case None =>
+          // we cannot differentiate between createPartyIfMissing or false, without breaking backwards compatibility
```
Comment (martinflorian-da) [inline apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/util/ValidatorUtil.scala]: Hm but why not do `createPartyIfMissing.getOrElse(true)` in this leg? This seems perfectly fine as a long-term solution to me...

It does mean that you need another if branch here though, and possibly another test... It's a fairly exotic edge condition but without this extra turn the semantics are a bit more confusing than they should be?

File: apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/admin/http/HttpValidatorHandler.scala (hunk 1 above)
Comment (martinflorian-da) [reply apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/admin/http/HttpValidatorHandler.scala]: I suggest picking `None` to preserve the original state as closely as possible; but that is bikeshedding on my part.

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/ValidatorIntegrationTest.scala (hunk 3 above)
Comment (pasindutennage-da) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/ValidatorIntegrationTest.scala:609]: Resolved: Added

File: apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/util/ValidatorUtil.scala (hunk 4 above)
Comment (pasindutennage-da) [reply apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/util/ValidatorUtil.scala]: Resolved: Covered all 4 possible cases

File: apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/admin/http/HttpValidatorHandler.scala (hunk 1 above)
Comment (pasindutennage-da) [reply apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/admin/http/HttpValidatorHandler.scala]: With the new strciter rules, we need this exatly as true

File: apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/util/ValidatorUtil.scala [hunk 5]
```diff
@@ -110,17 +111,66 @@ private[validator] object ValidatorUtil {
...
+          if (createPartyIfMissing.getOrElse(true)) {
+            logger.debug(s"No party ID provided and creation allowed. Allocating default party.")
+            connection
+              .getOrAllocateParty(
+                endUserName,
+                Seq(),
+                participantAdminConnection,
+              )
+          } else {
+
```
Comment (martinflorian-da) [inline apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/util/ValidatorUtil.scala]:
```suggestion
```

OCD nit

File: project/ignore-patterns/canton_network_test_log.ignore.txt [hunk 6]
```diff
@@ -118,6 +118,9 @@ Failed to connect to scan of .*DisasterRecoveryIntegrationTest
 Failed to connect to scan of .*AppUpgradeIntegrationTest
 Failed to connect to scan of .*ConfigurationProvidedBftScanConnectionIntegrationTest
 
+# Testing for a deliberately erroneous scenario where no party ID is provided
+party_id must be provided.*ValidatorIntegrationTest
```
Comment (martinflorian-da) [inline project/ignore-patterns/canton_network_test_log.ignore.txt]: Can't we add this ignore directly in the test via wrapping the problematic call in some variant of `assertLogs`?

File: docs/src/release_notes.rst [hunk 7]
```diff
@@ -7,6 +7,9 @@
 
 .. release-notes:: upcoming
 
+  - Validator
+    - Added support for specifying a custom party hint during user onboarding via the `/v0/admin/users` API.
```
Comment (martinflorian-da) [inline docs/src/release_notes.rst]:
```suggestion
    - Added support for picking a custom name for new parties created when onboarding users via the `/v0/admin/users` API.
```

Also not supper happy with that, but I found your version slightly misleading.

Comment (martinflorian-da) [reply docs/src/release_notes.rst]: Consider also adding a link to the updated docs page.

File: docs/src/validator_operator/validator_users.rst [hunk 8]
```diff
@@ -49,6 +55,18 @@ In order to associate a user with the party of the validator operator, the follo
...
+This is used to create parties with descriptive hints (e.g., treasury_dept::...) instead of OAuth IDs.
+
+1. Follow steps 1-3 from Mode ``Association with Existing Party`` to obtain the USER and TOKEN.
+2. Determine the desired, full PartyID ``(Hint::Namespace)``, e.g., ``alice::f3d917....``. Use the namespace of an existing party (like the validator operator's) for the namespace part. Save the full ID in a ``PARTY_ID`` environment variable: ``export PARTY_ID=alice::f3d917...``.
+3. Run the following command to create the new party and associate the user:
+
+.. code-block:: bash
+
+    curl -X POST -H "Authorization: Bearer $TOKEN" -H "Content-Type: application/json" \
+    --data-raw "{\"party_id\":\"$PARTY_ID\",\"name\":\"$USER\",\"createPartyIfMissing\":true}" \
```
Comment (martinflorian-da) [inline docs/src/validator_operator/validator_users.rst:67]: tbh I think I'd skip the extra vars and just used something like `<your party ID>` here; but no strong opinion

Comment (martinflorian-da) [reply docs/src/validator_operator/validator_users.rst:67]: ah nvm I see that you're just being consistent with the rest of the page; all perfect then

File: docs/src/validator_operator/validator_users.rst [hunk 9]
```diff
@@ -13,13 +13,19 @@ Canton distinguishes between parties and users, as documented in detail in the
 a party is an identity on-ledger, while a user represents an off-ledger entity
 that can be associated with one or more parties.
 
-By default, when a user logs in for the first time in the wallet, and presses the "Onboard yourself" button,
+The validator API endpoint `/v0/admin/users` supports three distinct modes for creating user and party associations:
```
Comment (martinflorian-da) [inline docs/src/validator_operator/validator_users.rst]: I'd move this a bit lower; many people really just need the UI button; no need to scare them with API endpoints?

File: docs/src/validator_operator/validator_users.rst [hunk 10]
```diff
@@ -13,13 +13,19 @@ Canton distinguishes between parties and users, as documented in detail in the
...
+
+1. Default Party Creation: Allocating a new party using the user ID as the party hint. The user is associated with this newly created party.
+2. Association with Existing Party: Linking a user to an existing party for wallet sharing.
+3. Custom Party Hint: Allocating a new party using a human-readable party hint. The user is associated with this newly created party.
```
Comment (martinflorian-da) [inline docs/src/validator_operator/validator_users.rst]: This seems out of sync with the code?

File: project/ignore-patterns/canton_network_test_log.ignore.txt (hunk 6 above)
Comment (pasindutennage-da) [reply project/ignore-patterns/canton_network_test_log.ignore.txt]: Done

File: docs/src/release_notes.rst (hunk 7 above)
Comment (pasindutennage-da) [reply docs/src/release_notes.rst]: Both done

File: docs/src/validator_operator/validator_users.rst (hunk 9 above)
Comment (pasindutennage-da) [reply docs/src/validator_operator/validator_users.rst]: Done

File: docs/src/validator_operator/validator_users.rst (hunk 10 above)
Comment (pasindutennage-da) [reply docs/src/validator_operator/validator_users.rst]: Rephrased, Done

File: apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/admin/http/HttpValidatorHandler.scala (hunk 1 above)
Comment (martinflorian-da) [reply apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/admin/http/HttpValidatorHandler.scala]: ...or `None`, which you do `getOrElse(true)` on...

Comment (pasindutennage-da) [reply apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/admin/http/HttpValidatorHandler.scala]: Done

Comment (pasindutennage-da) [reply apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/admin/http/HttpValidatorHandler.scala]: Done
